### PR TITLE
Additional tracing modes

### PR DESCRIPTION
**Pull request by pathtofile**
This PR add two additional tracing modes:
* 'all processes', to trace existing processes (except for tracee itself)
* 'all containers', to trace existing containers

To do this, the `--pid` and `--container` options altered, inline with discussions on https://github.com/aquasecurity/tracee/issues/221 

## pid
`--pid` has become `--pid-filter`, and has 3 options:
* `--pid-filter all`
* `--pid-filter new` (default)
* `--pid-filter id:<number>`, where `<number>` is a specific PID to trace
`id`  can be specified multiple times like the old `--pid`, but is mutually exlusive with `all` and `new`.
In the future this could be extended with `children:<number>` to only trace children of a PID. *NOTE* this would break existing scripts, so another possible solution is to replace `--pid-filter id:<number>` with just `--pid-filter <number>`

## container
`--container` has become `--container-filter`, and has 3 options:
* `--container-filter all`
* `--container-filter any` (previous behaviour of `--container`
The default is the flag not being set, which like the old `system` mode means all prosesses on the system and inside containers are traced.
In the future this could be extended with `id:<container_id>` to only trace a specific container

Apologies if the Go is Goarbage, my experience is more in C. Happy to split this up into multiple PRs if needed. Realted Issues: https://github.com/aquasecurity/tracee/issues/29 https://github.com/aquasecurity/tracee/issues/221 https://github.com/aquasecurity/tracee/issues/255

### CODE REVIEW

**Review by yanivagman — CHANGES_REQUESTED** [review]
This PR is very useful - good job!
Few small comments indside.

Let's finish the discussion (#221 ) before merging this

File: tracee/event_monitor_ebpf.c [hunk 1]
```diff
@@ -1181,10 +1197,11 @@ struct bpf_raw_tracepoint_args *ctx
     if (!should_trace())
         return 0;
 
+    // TODO PATH
```
[the remark this answers is not in the document]
Comment (pathtofile) [reply tracee/event_monitor_ebpf.c:1200]: dummy code, will remove

File: tracee/event_monitor_ebpf.c [hunk 2]
```diff
@@ -453,15 +455,28 @@ static __always_inline int get_config(u32 key)
...
+    else if (config_mode == MODE_CONTAINER_NEW)
         rc = lookup_pid_ns(task);
-    else
+    else if (config_mode == MODE_PROCESS_NEW || config_mode == MODE_PROCESS_LIST)
         rc = lookup_pid();
+    else if (config_mode == MODE_CONTAINER_ALL) {
+        // 'Container-All' means anything in a container
+        // We can check if we're in a namespace by checking
+        // our PID Vs 'real' PID on host
+        if (get_task_ns_tgid(task) != host_pid) {
```
Comment (pathtofile) [inline tracee/event_monitor_ebpf.c:476]: @yanivagman I draw your attention to this semi-hacky way to check if a process is running in a container or not, as `MODE_CONTAINER_ALL` really means 'any process thats in a container'. Are you ok with this assumtion?

File: main.go [hunk 3]
```diff
@@ -23,21 +24,65 @@ func main() {
...
+				// support specific container IDs
+				if c.String("container-filter") == "all" {
+					mode = tracee.ModeContainerAll
+				} else if c.String("container-filter") == "new" {
+					mode = tracee.ModeContainerNew
+				} else {
+					return fmt.Errorf("Invalid container-filter: %s", c.String("container-filter"))
+				}
+			} else if c.IsSet("pid-filter") {
+				hasExlusiveOption := false
```
[the remark this answers is not in the document]
Comment (pathtofile) [reply main.go]: oops, nice catch, will fix name

File: main.go [hunk 4]
```diff
@@ -23,21 +24,65 @@ func main() {
...
-			}
 			events, err := prepareEventsToTrace(c.StringSlice("event"), c.StringSlice("events-set"), c.StringSlice("exclude-event"))
 			if err != nil {
 				return err
 			}
+
+			// Set Default mode
+			mode := tracee.ModeProcessNew
+			var pidsToTrace []int
+			if c.IsSet("container-filter") && c.IsSet("pid-filter") {
```
[the remark this answers is not in the document]
Comment (pathtofile) [reply main.go:35]: Sure, I'll change `container-filter` to `container` and `process-filter` to `process`

File: main.go [hunk 5]
```diff
@@ -104,17 +149,17 @@ func main() {
...
 				Usage:   "just list tracable events",
 			},
-			&cli.BoolFlag{
-				Name:    "container",
+			&cli.StringFlag{
+				Name:    "container-filter",
 				Aliases: []string{"c"},
-				Value:   false,
-				Usage:   "trace only containers",
+				Value:   "",
```
[the remark this answers is not in the document]
Comment (pathtofile) [reply main.go:155]: Sorry, I'm reading the docs for the `cli` package and I don't think I understand - what's the use of Value in flags?

The original default-default begaviour when you just run `./tracee` is to do all new system processes, correct? does setting the container default value to `new` then mean running `./tracee` makes it by default trace all new container-only processes?

File: main.go [hunk 6]
```diff
@@ -23,21 +24,65 @@ func main() {
...
+						hasExlusiveOption = true
+						if hasIDOption {
+							return fmt.Errorf("pid-filter 'all', 'new', and 'id:<number> are all mutually exclusive")
+						}
+						if pidFilter == "all" {
+							mode = tracee.ModeProcessAll
+						} else {
+							mode = tracee.ModeProcessNew
+						}
+					} else if pidFilter[:3] == "id:" {
```
[the remark this answers is not in the document]
Comment (pathtofile) [reply main.go:64]: Yep sure makes sense, that will also make sure everyone elses scripts that just used `--pid` don't need to change, another advantage

File: main.go [hunk 7]
```diff
@@ -104,17 +149,17 @@ func main() {
...
+				Usage:   "container filtering setting. Options are 'new' to trace only new containers; 'all' to trace all presesses inside containers.",
 			},
-			&cli.IntSliceFlag{
-				Name:    "pid",
+			&cli.StringSliceFlag{
+				Name:    "pid-filter",
 				Aliases: []string{"p"},
 				Value:   nil,
-				Usage:   "trace only the specified pid. use this flag multiple times to choose multiple pids",
+				Usage:   "pid filtering setting. Options are: 'new' to trace only new pids (default); 'all' to trace all pids; 'id:<number>' to trace only the specified pid. Use 'id' flag multiple times to choose multiple pids",
```
[the remark this answers is not in the document]
Comment (pathtofile) [reply main.go:162]: agreed

**Comment by pathtofile** [comment]
Fixed initial round of review while we discuss https://github.com/aquasecurity/tracee/issues/221

File: tracee/event_monitor_ebpf.c (hunk 2 above)
Comment (yanivagman) [reply tracee/event_monitor_ebpf.c:476]: Yes, I'm ok with it. 
As kernel isn't aware about containers, but rather works with namespaces and cgroups, there is no other way than using these kind of hacks. For example, I used "pid == 1" in the code to know if a new container started (the assumption here is that a container will use a new pid namespace, which is not always the case).

In kernel 4.18, new bpf helpers were added that allow to better work with cgroups (e.g. bpf_get_current_cgroup_id()) - we may change the way we identify containers in the future to work with these helpers

File: main.go [hunk 8]
```diff
@@ -104,17 +147,17 @@ func main() {
 				Value:   false,
 				Usage:   "just list tracable events",
 			},
-			&cli.BoolFlag{
+			&cli.StringFlag{
 				Name:    "container",
 				Aliases: []string{"c"},
-				Value:   false,
-				Usage:   "trace only containers",
+				Value:   "new",
```
Comment (yanivagman) [inline main.go:153]: Value is indeed the default value, and we don't want to set to "new" here.
I wrote the comment here, as it seemed as the most suitable place - sorry for the confusion.

We need to keep the current default behavior (tracing new processes). What I meant here, is that if someone uses "--container" without providing any additional command (e.g. all/new) - we need to fall back to the default container mode behavior that exists today, which is tracing newly created containers.

I think that you can use "IsSet" for this purpose in addition to checking that the given string is an empty string

**Comment by pathtofile** [comment]
@yanivagman This is read for another look, I've implemented that `--trace` changes from the Discussion

File: main.go [hunk 9]
```diff
@@ -175,6 +172,67 @@ func main() {
...
+	// Convert to Traceing Mode
+	if traceType == "p" {
+		if traceOption == "all" {
+			mode = tracee.ModeProcessAll
+		} else if traceOption == "new" {
+			mode = tracee.ModeProcessNew
+		} else {
+			mode = tracee.ModeProcessList
+			// Attempt to split into PIDs
+			for _, pidString := range strings.Split(traceOption, ",") {
```
Comment (yanivagman) [inline main.go:215]: What about the case that someone gives "process:"?
In that case, traceOption is an empty string - and tracee will trace nothing

**Review by yanivagman — CHANGES_REQUESTED** [review]
Great.
Only had two small comments inline.

Just to be sure - did you test the different modes and verified they all work as expected?
One more thing so I can approve and merge - please merge all commits into one commit.
Thanks!

File: main.go [hunk 10]
```diff
@@ -175,6 +172,67 @@ func main() {
 	}
 }
 
+func prepareTraceMode(traceString string) (uint32, []int, error) {
+	// Set Default mode - all new Processes only
+	mode := tracee.ModeProcessNew
+	var pidsToTrace []int
+	traceHelp := "--trace can be the following options:\n"
+	traceHelp += "<blank> or 'p' or 'process' or 'process:new' | Trace new processes\n"
```
[the remark this answers is not in the document]
Comment (pathtofile) [reply main.go:180]: Nope I agree, thinking about it `--trace` doesn't really make sense to be a valid option, so removed references to it in the readme and code

File: main.go (hunk 9 above)
Comment (pathtofile) [reply main.go:215]: Good catch, will fix

**Comment by pathtofile** [comment]
Ah fuck sorry, got a newborn kiddo and my brain isn't functioning, I'm blowing away this messed up branch and will create a new PR with references to this, sorry

**Comment by yanivagman** [comment]
@pathtofile no worries, congratulations for the newborn :-)

**Pull request closed without merging**